redbot/core/commands/converters.py

```diff
@@ -0,0 +1,74 @@
+import re
+from datetime import timedelta
+from . import Converter, Context, BadArgument
+
+__all__ = ["TimeConverter"]
+
+
+class TimeConverter(Converter):
+    """
+    Convert to a :class:`datetime.timedelta` class.
+
+    The converter supports seconds, minutes, hours and days.
+
+    Time can be attached to the unit (only use the first letter) or not
+    (then use the full word).
+
+    .. admonition:: Example
+
+        *   ``50s`` = ``50 seconds`` -> :class:`datetime.timedelta(seconds=50)`
+        *   ``12m`` = ``12 minutes`` -> :class:`datetime.timedelta(seconds=720)`
+        *   ``4h`` = ``4 hours`` -> :class:`datetime.timedelta(seconds=14400)`
+        *   ``1d`` = ``1 day`` -> :class:`datetime.timedelta(days=1)`
+
+        Using the converter with a command:
+
+        .. code-block:: python3
+
+            @commands.command()
+            async def timer(self, ctx, time: commands.Time):
+                await self.start_timer(time)
+
+        Using the converter manually:
+
+        .. code-block:: python3
+
+            async def convert_time(ctx: Context, text: str) -> datetime.timedelta:
+                time = await commands.Time.convert(ctx, text)
+                return time
+    """
+
+    async def convert(ctx: Context, argument: str) -> timedelta:
+        """
+        Convert manually a string to a :class:`datetime.timedelta` class.
+        
+        .. warning:: This should not be called as a command function annotation.
+            This is for manual calls.
+
+        Arguments
+        ---------
+        ctx: Context
+            The context of the command.
+        argument: str
+            The string you want to convert.
+
+        Returns
+        -------
+        datetime.timedelta
+            The :class:`datetime.timedelta` object.
+
+        Raises
+        ------
+        ~discord.ext.commands.BadArgument
+            No time was found from the given string.
+        """
+        TIME_RE = re.compile(
+            r"((?P<days>\d+?)\s?(d(ays?)?))?\s?((?P<hours>\d+?)\s?(hours?|hrs|hr?))?\s?((?P<minutes>\d+?)\s?(minutes?|mins?|m))?\s?((?P<seconds>\d+?)\s?(seconds?|secs?|s))?\s?",
+            re.I,
+        )
+        matches = TIME_RE.match(argument)
+        params = {k: int(v) for k, v in matches.groupdict().items() if None not in (k, v)}
+        time = timedelta(**params)
+        if str(time) == "0:00:00":
```

Comment:
This disallows a zero second timedelta, which could be allowed by checking against `if not params` earlier in the code instead.

Reply:
But if the user sets, for example `2 weeks`, Regex will return 0:00:00, while the user is actually expecting something else. For example, with timed mute, if the user input a wrong unit, what will happen is that the user will get muted and immediately unmuted.

I'm trying to follow `discord.py`'s style for converters, and when it can't convert, it raises `BadArgument` (instead of returning None or something like that).

Reply:
Your handling of this places `0 seconds` as invalid, when the intent is to not handle `2 weeks`. Meanwhile, if `params` is an empty dict here, you can raise the bad argument there (having matched nothing expected)